### ISSUE TITLE
Average Rank Excluding Self calculation not intuitive to users

### ISSUE THREAD

**Issue by wenmogu** [TEAMMATES/teammates#8280]
**Description of feature/enhancement**
The `Average Rank Excluding Self` is calculated from the normalized ranks of the ranks which have already excluded the self-ranks. However, in my opinion, it makes more sense to use the normalized ranks of ranks which haven't excluded the self-ranks to calculate the average.

The example below is formatted into excel sheet from the expected result of JUnit test testing method `getFeedbackSessionResultsSummaryAsCsv` in class `FeedbackSessionsLogicTest`. Inside the summary, the `Average Rank Excluding Self` in the selected line is calculated to be 1(Credit to @joanneong who discovered this unexpected number). We should expect the number to be calculated as `(the sum of the ranks received - self rank) / the number of non-self ranks received`, which results in 1.5, instead of 1.
![image](https://user-images.githubusercontent.com/28526038/34516922-d6af55a0-f0b3-11e7-8653-eb660fad830b.png)

The steps below outlines the calculation done in the current implementation:
1. extract the non-self-ranks from the raw data, which means, out of the 8 ranks given in the responses, the 4 self-ranks are ignored
2. normalize these 4 non-self-ranks. This means the unrated are ignored and ties are resolved.
E.g. given the two responses below (the arrays used below are only for illustration purpose),
`[giver=student3, recipient=student4, answer=4]`, and
`[giver=student3, recipient=student3, answer=3]`, 
if there is no removal of the self-rank, the normalized result will become
`[giver=student3, recipient=student4, answer=4, rankAfterNormalization=2]`, and
`[giver=student3, recipient=student3, answer=3, rankAfterNormalization=1]`.
Therefore, if the removal is done before the normalization, only the non-self-rank `[giver=student3, recipient=student4, answer=4]` will be normalized and resulting in `[giver=student3, recipient=student4, answer=4, rankAfterNormalization=1]`, which in turn leads to the `Average Rank Excluding Self = 1` instead of 1.5.

**Justification**

As explained above, the calculation for `Average Rank Excluding Self` is not intuitive enough for the users and there is no mention of it in documentation. Hence it will confuse the users and get inaccurate feedback results.

**Comment by joanneong** [comment]
In fact, the entire ranking system is not quite intuitive to users. This is particularly because of the normalization mechanism that @wenmogu elaborated above, which is not well explained in the current documentation as the current documentation within the instructor help page only mentions what happens in the event of ties for rank questions. (@wenmogu and I spent a fair bit of time puzzling out how the data in the excel screenshot above came about!)

 It might be better if, similar to how the contribution calculation scheme is explained [here](https://docs.google.com/document/d/1hjQQHYM3YId0EUSrGnJWG5AeFpDD_G7xg_d--7jg3vU/pub?embedded=true#h.n5u2xs6z9y0g), the entire calculation mechanism behind rank questions (normalization and calculations excluding self) is also made transparent to interested users (both students and instructors).

In short, there should minimally be improved documentation for rank questions. However, I believe that @wenmogu is also further suggesting an alternative ranking scheme with respect to calculating `Average Rank Excluding Self`.

**Comment by damithc** [comment]
Thanks for raising this @wenmogu and for inputs @joanneong 
_Average excluding self_ should be calculated as if self-ranking was not done at all.
On a related note, I wonder if we should give 'Overall ranking' and 'Overall ranking (excluding self-ranking)' instead of (or in addition to) average values. I think average values don't make much sense as output when the input was rankings.

**Comment by mrkagelui** [comment]
I can take it up if modification or implementation of the logic is required. (I tried to read through the discussion but it seems a lot of background to pick up here 😆 )

**Comment by wenmogu** [comment]
Sorry for the reply. Here are some of my thoughts on this issue: 
1) There should not be _average_ ranks as the existence does not make sense. Although the current _Average excluding self_ serves as an indicator of how others rate one's performance, the naming/computation of this value has to be changed.
2) I think "Overall ranking" and "Overall ranking (excluding self-ranking)" make a lot more sense than _Average excluding self_. I feel that overall rankings give a conclusion to the raw data and can be used in the final evaluation of the students directly. There are several ways of computing an overall ranking, and I think the two below are easy to understand and implement. 
    - Compute the overall ranking like how the countries are ranked on [Olympiad (I mean the sports Olympiad) medal lists](https://en.wikipedia.org/wiki/Olympic_medal_table#Medal_count_ranking): if A's number of Rank 1 received > B's number of Rank 1 received, then A is ranked before B; else, compare the number of Rank 2 received, if equal then compare the number of Rank 3 received and so on. 
    - Assign marks to each rank received, e.g. 100 points for Rank 1, 75 points for Rank 2, 50 points for Rank 3 and 25 points for Rank 4. Then the overall rank can be computed by ranking the total points that the students receive. 

And to be honest, both @joanneong and I feel that the documentation regarding the rank questions needs to be improved. Currently the example given on Instructor Help page is inadequate (as compared to the better example in the comment of the normalization function in the code (as seen below), Moreover, the explanation of the working is not explained and that might confuse the students/instructors. 

The example given on instructor help page regarding normalization of ranks: 
![image](https://user-images.githubusercontent.com/28526038/35179169-5a0b5fa2-fdcf-11e7-9c4f-6ba65e87e8c4.png)

The example given in the code: 
![image](https://user-images.githubusercontent.com/28526038/35179150-13aa1c42-fdcf-11e7-8672-00d36c6f939b.png)

**Comment by damithc** [comment]
The 2nd way of calculation may be easier to explain. The first algorithm (or something similar) can be used to break ties resulting from the second algorithm.

Yes, documentation can be improved for sure.

Perhaps one of you (or both together) can take this up? @joanneong @wenmogu

**Comment by joanneong** [comment]
Got it. @wenmogu and I will both be trying to work on this right now. We will come to a consensus on the ranking system, and then split the work thereafter. 

Unfortunately, the documentation will probably be completed before any new ranking system is fully implemented, but not be merged until after the new ranking system is completed.

**Comment by joanneong** [comment]
After some discussion, here is an update on what we will be working on for this particular issue. Note that the discussion here is limited to rank recipients questions only, and excludes rank options questions since rank options questions do not suffer from the same issue.

**(1) The revamped ranking system**
We intend to retain the normalization system, but change the computation that follows after the normalization.

The original system currently provides the following results:
![image](https://user-images.githubusercontent.com/24493463/35179750-4610146e-fddb-11e7-9cc0-5774381c4030.png)

We propose to change it to:
![image](https://user-images.githubusercontent.com/24493463/35179752-5c413e7a-fddb-11e7-95ea-2b62b72f541a.png)

We intend to adopt a weighted ranking system, whereby the lowest rank is worth 25 points and each higher rank is worth 25 points more than the succeeding rank. For instance, if the question requires a student to rank 4 recipients, the points will be: rank 4 = 25; rank 3 = 50; rank 2 = 75; rank 1 = 100.

The total points will then be tabulated, and average points will be calculated as well.

If a person is not rated, then, as per the original system, the person's details will not be displayed for the instructor.

**(2) To do**
With the system decided upon, we have come up with a simplified list of the things to do in order to fix this issue. 
a) Documentation (instructor help page): update the documentation involving rank recipients questions to explain the entire ranking system (normalization and points system)
b) Logic: Set up the revamped ranking system as explained above
c) UI: Set up the display as shown in the mock-up above

I will be starting from the documentation while @wenmogu will start from the logic. After I am done with the documentation (which acts like a scaffold for the logic), I will help @wenmogu with the logic as well.

**Comment by damithc** [comment]
As discussed, let's not add a point system which may be a bit surprising to users. Instead, let's give an overall ranking (including self, and excluding self).
